pkg/datastore/test/caveat.go

```diff
@@ -0,0 +1,298 @@
+	req.Equal(oldRev, fetchedRev)
+}
+
+func CaveatedRelationshipWatchTest(t *testing.T, tester DatastoreTester) {
+	req := require.New(t)
+	ds, err := tester.New(0*time.Second, veryLargeGCWindow, 16)
+	req.NoError(err)
+
+	skipIfNotCaveatStorer(t, ds)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Write caveat and caveated relationship
+	coreCaveat := createCoreCaveat(t)
+	_, err = writeCaveat(ctx, ds, coreCaveat)
+	req.NoError(err)
+
+	// TODO bug: Watch API won't send updates i revision used is the first revision
+	lowestRevision, err := ds.HeadRevision(ctx)
+	req.NoError(err)
+	chanRevisionChanges, chanErr := ds.Watch(ctx, lowestRevision)
+	req.Zero(len(chanErr))
+
+	tpl := createTestCaveatedTuple(t, "document:companyplan#parent@folder:company#...", coreCaveat.Name)
+	_, err = common.WriteTuples(ctx, ds, core.RelationTupleUpdate_CREATE, tpl)
+	req.NoError(err)
+
+	// Caveated Relationship should come through Watch API
+
+	changeWait := time.NewTimer(5 * time.Second)
+	select {
+	case change, ok := <-chanRevisionChanges:
+		req.True(ok)
+		req.Len(change.Changes, 1)
+		for _, update := range change.Changes {
+			foundDiff := cmp.Diff(tpl, update.Tuple, protocmp.Transform())
+			req.Empty(foundDiff)
+		}
+	case <-changeWait.C:
+		req.Fail("timed out waiting for caveated relationship via Watch API")
+	}
+}
+
+func skipIfNotCaveatStorer(t *testing.T, ds datastore.Datastore) {
+	ctx := context.Background()
+	_, _ = ds.ReadWriteTx(ctx, func(ctx context.Context, transaction datastore.ReadWriteTransaction) error { //nolint: errcheck
+		_, _, err := transaction.ReadCaveatByName(ctx, uuid.NewString())
+		if !errors.As(err, &datastore.ErrCaveatNameNotFound{}) {
+			t.Skip("datastore does not implement CaveatStorer interface")
+		}
+		return fmt.Errorf("force rollback of unnecesary tx")
+	})
+}
+
+func createTestCaveatedTuple(t *testing.T, tplString string, caveatName string) *core.RelationTuple {
+	tpl := tuple.MustParse(tplString)
+	st, err := structpb.NewStruct(map[string]interface{}{"a": 1, "b": "test"})
+	require.NoError(t, err)
+
+	tpl.Caveat = &core.ContextualizedCaveat{
+		CaveatName: caveatName,
+		Context:    st,
+	}
+	return tpl
+}
+
+func writeCaveats(ctx context.Context, ds datastore.Datastore, coreCaveat ...*core.Caveat) (datastore.Revision, error) {
+	rev, err := ds.ReadWriteTx(ctx, func(ctx context.Context, tx datastore.ReadWriteTransaction) error {
+		return tx.WriteCaveats(coreCaveat)
+	})
+	if err != nil {
+		return datastore.NoRevision, err
+	}
+	return rev, err
+}
+
+func writeCaveat(ctx context.Context, ds datastore.Datastore, coreCaveat *core.Caveat) (datastore.Revision, error) {
+	rev, err := writeCaveats(ctx, ds, coreCaveat)
+	if err != nil {
+		return datastore.NoRevision, err
+	}
+	return rev, nil
+}
+
+func createCoreCaveat(t *testing.T) *core.Caveat {
+	t.Helper()
+	c := createCompiledCaveat(t)
+	cBytes, err := c.Serialize()
+	require.NoError(t, err)
+
+	coreCaveat := &core.Caveat{
```

Comment:
Might also be worth renaming this type to `CaveatDefinition` to match `NamespaceDefinition`

Reply:
I can do in follow up to, sounds good!